File: backend-springboot/src/main/java/com/vaultx/banking/entity/Customer.java
package com.vaultx.banking.entity;

import jakarta.persistence.*;
import jakarta.validation.constraints.Email;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;

@Entity
@Table(name = "customers")
public class Customer {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotBlank
    @Column(name = "first_name")
    private String firstName;

    @NotBlank
    @Column(name = "last_name")
    private String lastName;

    @Email
    @NotBlank
    @Column(unique = true)
    private String email;

    @NotBlank
    private String phone;

    @Column(name = "date_of_birth")
    private LocalDate dateOfBirth;

    @NotBlank
    private String password;

    @Column(name = "account_number", unique = true)
    private String accountNumber;

    @Enumerated(EnumType.STRING)
    @Column(name = "account_type")
    private AccountType accountType;

    @Column(name = "initial_deposit", precision = 15, scale = 2)
    private BigDecimal initialDeposit;

    @Column(name = "current_balance", precision = 15, scale = 2)
    private BigDecimal currentBalance;

    @Column(name = "pan_number")
    private String panNumber;

    @Column(name = "aadhar_number")
    private String aadharNumber;

    private String occupation;

    @Column(name = "monthly_income")
    private String monthlyIncome;

    @Enumerated(EnumType.STRING)
    private CustomerStatus status;

    @Column(name = "kyc_completed")
    private boolean kycCompleted;

    @Column(name = "can_perform_operations")
    private boolean canPerformOperations;

    @Column(name = "submitted_date")
    private LocalDate submittedDate;

    @Column(name = "kyc_date")
    private LocalDate kycDate;

    @Column(name = "approval_date")
    private LocalDate approvalDate;

    @Column(name = "clerk_notes", columnDefinition = "TEXT")
    private String clerkNotes;

    @Column(name = "manager_notes", columnDefinition = "TEXT")
    private String managerNotes;

    @Column(name = "created_at")
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    public enum AccountType {
        SAVINGS, CURRENT
    }

    public enum CustomerStatus {
        DOCUMENTS_SUBMITTED,
        KYC_IN_PROGRESS,
        KYC_COMPLETED,
        APPROVED,
        REJECTED
    }

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
        if (submittedDate == null) {
            submittedDate = LocalDate.now();
        }
        if (status == null) {
            status = CustomerStatus.DOCUMENTS_SUBMITTED;
        }
        if (currentBalance == null) {
            currentBalance = initialDeposit != null ? initialDeposit : BigDecimal.ZERO;
        }
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }

    // Constructors
    public Customer() {}

    public Customer(String firstName, String lastName, String email, String phone, String password) {
        this.firstName = firstName;
        this.lastName = lastName;
        this.email = email;
        this.phone = phone;
        this.password = password;
    }

    // Getters and Setters
    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }

    public String getFirstName() { return firstName; }
    public void setFirstName(String firstName) { this.firstName = firstName; }

    public String getLastName() { return lastName; }
    public void setLastName(String lastName) { this.lastName = lastName; }

    public String getEmail() { return email; }
    public void setEmail(String email) { this.email = email; }

    public String getPhone() { return phone; }
    public void setPhone(String phone) { this.phone = phone; }

    public LocalDate getDateOfBirth() { return dateOfBirth; }
    public void setDateOfBirth(LocalDate dateOfBirth) { this.dateOfBirth = dateOfBirth; }

    public String getPassword() { return password; }
    public void setPassword(String password) { this.password = password; }

    public String getAccountNumber() { return accountNumber; }
    public void setAccountNumber(String accountNumber) { this.accountNumber = accountNumber; }

    public AccountType getAccountType() { return accountType; }
    public void setAccountType(AccountType accountType) { this.accountType = accountType; }

    public BigDecimal getInitialDeposit() { return initialDeposit; }
    public void setInitialDeposit(BigDecimal initialDeposit) { this.initialDeposit = initialDeposit; }

    public BigDecimal getCurrentBalance() { return currentBalance; }
    public void setCurrentBalance(BigDecimal currentBalance) { this.currentBalance = currentBalance; }

    public String getPanNumber() { return panNumber; }
    public void setPanNumber(String panNumber) { this.panNumber = panNumber; }

    public String getAadharNumber() { return aadharNumber; }
    public void setAadharNumber(String aadharNumber) { this.aadharNumber = aadharNumber; }

    public String getOccupation() { return occupation; }
    public void setOccupation(String occupation) { this.occupation = occupation; }

    public String getMonthlyIncome() { return monthlyIncome; }
    public void setMonthlyIncome(String monthlyIncome) { this.monthlyIncome = monthlyIncome; }

    public CustomerStatus getStatus() { return status; }
    public void setStatus(CustomerStatus status) { this.status = status; }

    public boolean isKycCompleted() { return kycCompleted; }
    public void setKycCompleted(boolean kycCompleted) { this.kycCompleted = kycCompleted; }

    public boolean isCanPerformOperations() { return canPerformOperations; }
    public void setCanPerformOperations(boolean canPerformOperations) { this.canPerformOperations = canPerformOperations; }

    public LocalDate getSubmittedDate() { return submittedDate; }
    public void setSubmittedDate(LocalDate submittedDate) { this.submittedDate = submittedDate; }

    public LocalDate getKycDate() { return kycDate; }
    public void setKycDate(LocalDate kycDate) { this.kycDate = kycDate; }

    public LocalDate getApprovalDate() { return approvalDate; }
    public void setApprovalDate(LocalDate approvalDate) { this.approvalDate = approvalDate; }

    public String getClerkNotes() { return clerkNotes; }
    public void setClerkNotes(String clerkNotes) { this.clerkNotes = clerkNotes; }

    public String getManagerNotes() { return managerNotes; }
    public void setManagerNotes(String managerNotes) { this.managerNotes = managerNotes; }

    public LocalDateTime getCreatedAt() { return createdAt; }
    public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }

    public LocalDateTime getUpdatedAt() { return updatedAt; }
    public void setUpdatedAt(LocalDateTime updatedAt) { this.updatedAt = updatedAt; }
}

File: backend-dotnet/VaultX.Banking.API/Models/Customer.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace VaultX.Banking.API.Models;

public class Customer
{
    [Key]
    public int Id { get; set; }

    [Required]
    [StringLength(100)]
    public string FirstName { get; set; } = string.Empty;

    [Required]
    [StringLength(100)]
    public string LastName { get; set; } = string.Empty;

    [Required]
    [EmailAddress]
    [StringLength(255)]
    public string Email { get; set; } = string.Empty;

    [Required]
    [StringLength(20)]
    public string Phone { get; set; } = string.Empty;

    public DateOnly? DateOfBirth { get; set; }

    [Required]
    [StringLength(255)]
    public string Password { get; set; } = string.Empty;

    [StringLength(50)]
    public string? AccountNumber { get; set; }

    public AccountType AccountType { get; set; }

    [Column(TypeName = "decimal(15,2)")]
    public decimal InitialDeposit { get; set; }

    [Column(TypeName = "decimal(15,2)")]
    public decimal CurrentBalance { get; set; }

    [StringLength(10)]
    public string? PanNumber { get; set; }

    [StringLength(12)]
    public string? AadharNumber { get; set; }

    [StringLength(100)]
    public string? Occupation { get; set; }

    [StringLength(50)]
    public string? MonthlyIncome { get; set; }

    public CustomerStatus Status { get; set; } = CustomerStatus.DocumentsSubmitted;

    public bool KycCompleted { get; set; } = false;

    public bool CanPerformOperations { get; set; } = false;

    public DateOnly SubmittedDate { get; set; } = DateOnly.FromDateTime(DateTime.Now);

    public DateOnly? KycDate { get; set; }

    public DateOnly? ApprovalDate { get; set; }

    public string? ClerkNotes { get; set; }

    public string? ManagerNotes { get; set; }

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;

    // Navigation property
    public virtual ICollection<Transaction> Transactions { get; set; } = new List<Transaction>();
}

public enum AccountType
{
    Savings,
    Current
}

public enum CustomerStatus
{
    DocumentsSubmitted,
    KycInProgress,
    KycCompleted,
    Approved,
    Rejected
}
